### PR TITLE
Expose additional functions of the crypto engine to user

### PR DESCRIPTION
**Pull request by twisted-pear**
# What's new

- Allow use of the standard `furi_hal_crypto_encrypt` and `furi_hal_crypto_decrypt` functions with user supplied keys instead of using keys from the secure enclave. User supplied key can be loaded with `furi_hal_crypto_load_key` and unloaded with `furi_hal_crypto_unload_key`.
- Using user supplied keys, also allow use of the crypto engine's CTR and GCM mode. This can be done using `furi_hal_crypto_ctr` and `furi_hal_crypto_gcm`. For GCM there are convenience functions provided for encryption and decryption (`furi_hal_crypto_gcm_encrypt_and_tag` and `furi_hal_crypto_gcm_decrypt_and_verify`).
- Introduce a define called `FURI_HAL_CRYPTO_ADVANCED_AVAIL` which indicates the availability of these new functions.
- This is somewhat related to Issue #2885. It is not intended to affect the secure enclave in any way.

# Verification 

- Use of `furi_hal_crypto_encrypt` and `furi_hal_crypto_decrypt` remains unchanged. Data encrypted using a user supplied key and `furi_hal_crypto_encrypt` can be decrypted using the same key and `furi_hal_crypto_decrypt`.
- The output of the CTR function is intended to be identical to the output generated by the openssl CLI tool.
- The output of the GCM function is intended to be identical to the output generated by the GCM functions in this repo: https://github.com/mko-x/SharedAES-GCM
- I wrote a FAP automating all these tests which is available here: https://github.com/twisted-pear/crypto_test. This app will test CTR and GCM with a few preconfigured test vectors. It writes the results to the log at level Info.

# Checklist (For Reviewer)

- [x] PR has description of feature/bug or link to Confluence/Jira task
- [x] Description contains actions to verify feature/bugfix
- [x] I've built this code, uploaded it to the device and verified feature/bugfix

### CODE REVIEW

**Comment by twisted-pear** [comment]
Updated the API version and symbols this time so the automated builds can run through

**Comment by twisted-pear** [comment]
How do I get around this:
```
API versions aren't matching for available targets. Please update!
==> firmware/targets/f18/api_symbols.csv <==
entry,status,name,type,params
Version,+,34.3,,

==> firmware/targets/f7/api_symbols.csv <==
entry,status,name,type,params
Version,+,34.4,,
```
Can I build for f18 myself somehow? Should I just update the api_symbols.csv manually?

**Comment by hedger** [comment]
@twisted-pear just run `./fbt TARGET_HW=18`.

**Comment by twisted-pear** [comment]
thanks, i already figured that out
should build successfully now

File: firmware/targets/f7/furi_hal/furi_hal_crypto.c [hunk 1]
```diff
@@ -276,6 +289,35 @@ bool furi_hal_crypto_store_unload_key(uint8_t slot) {
     return (shci_state == SHCI_Success);
 }
 
+bool furi_hal_crypto_load_key(const uint8_t* key, const uint8_t* iv) {
+    furi_assert(furi_hal_crypto_mutex);
+    furi_check(furi_mutex_acquire(furi_hal_crypto_mutex, FuriWaitForever) == FuriStatusOk);
+
+    furi_hal_bus_enable(FuriHalBusAES1);
```
Comment (skotopes) [inline firmware/targets/f7/furi_hal/furi_hal_crypto.c:301]: I'd set AES_CR_EN and never ever touch it till key unloaded.

File: firmware/targets/f7/furi_hal/furi_hal_crypto.c [hunk 2]
```diff
@@ -276,6 +289,35 @@ bool furi_hal_crypto_store_unload_key(uint8_t slot) {
...
+
+    if(!furi_hal_bt_is_alive()) {
```
Comment (skotopes) [inline firmware/targets/f7/furi_hal/furi_hal_crypto.c]: Only if enclave used

File: firmware/targets/f7/furi_hal/furi_hal_crypto.c [hunk 3]
```diff
@@ -340,3 +382,377 @@ bool furi_hal_crypto_decrypt(const uint8_t* input, uint8_t* output, size_t size)
...
+}
+
+static bool
+    furi_hal_crypto_load_key_bswap(const uint8_t* key, const uint8_t* iv, uint32_t chaining_mode) {
+    furi_assert(furi_hal_crypto_mutex);
+    furi_check(furi_mutex_acquire(furi_hal_crypto_mutex, FuriWaitForever) == FuriStatusOk);
+
+    furi_hal_bus_enable(FuriHalBusAES1);
+
+    if(!furi_hal_bt_is_alive()) {
```
Comment (skotopes) [inline firmware/targets/f7/furi_hal/furi_hal_crypto.c]: same

File: firmware/targets/f7/furi_hal/furi_hal_crypto.c [hunk 4]
```diff
@@ -340,3 +382,377 @@ bool furi_hal_crypto_decrypt(const uint8_t* input, uint8_t* output, size_t size)
...
+    SET_BIT(AES1->CR, AES_CR_EN);
+    MODIFY_REG(AES1->CR, AES_CR_MODE, CRYPTO_MODE_ENCRYPT);
+
+    bool state = false;
+    size_t last_block_bytes = length % CRYPTO_BLK_LEN;
+
+    size_t i;
+    for(i = 0; i < length - last_block_bytes; i += CRYPTO_BLK_LEN) {
+        if(!furi_hal_crypto_process_block_bswap(&input[i], &output[i], CRYPTO_BLK_LEN)) {
+            goto out_enc_err;
```
Comment (skotopes) [inline firmware/targets/f7/furi_hal/furi_hal_crypto.c]: We prohibit use of `goto`, instead we use `do{ }while(0)`.Moment when you see issue with do-while usually means that complexity of your code is not acceptable anymore.

File: firmware/targets/f7/furi_hal/furi_hal_crypto.c [hunk 5]
```diff
@@ -340,3 +382,377 @@ bool furi_hal_crypto_decrypt(const uint8_t* input, uint8_t* output, size_t size)
...
+    }
+
+    return true;
+}
+
+static bool furi_hal_crypto_gcm_header(const uint8_t* aad, size_t aad_length) {
+    /* GCM header phase */
+
+    MODIFY_REG(AES1->CR, AES_CR_GCMPH, CRYPTO_GCM_PH_HEADER);
+    SET_BIT(AES1->CR, AES_CR_EN);
```
Comment (skotopes) [inline firmware/targets/f7/furi_hal/furi_hal_crypto.c:552]: looks like you do want to make mode change method that disables CR_EN, changes mode, and then re-enables CR_EN

File: firmware/targets/f7/furi_hal/furi_hal_crypto.c [hunk 6]
```diff
@@ -340,3 +382,377 @@ bool furi_hal_crypto_decrypt(const uint8_t* input, uint8_t* output, size_t size)
...
+
+    if(last_block_bytes > 0) {
+        if(!furi_hal_crypto_process_block_no_read_bswap(&aad[i], last_block_bytes)) {
+            goto out_enc_err;
+        }
+    }
+
+    return true;
+
+out_enc_err:
```
Comment (skotopes) [inline firmware/targets/f7/furi_hal/furi_hal_crypto.c]: Nope

File: firmware/targets/f7/furi_hal/furi_hal_crypto.c [hunk 7]
```diff
@@ -340,3 +382,377 @@ bool furi_hal_crypto_decrypt(const uint8_t* input, uint8_t* output, size_t size)
...
+                AES1->CR, AES_CR_NPBLB, (CRYPTO_BLK_LEN - last_block_bytes) << AES_CR_NPBLB_Pos);
+        }
+        if(!furi_hal_crypto_process_block_bswap(&input[i], &output[i], last_block_bytes)) {
+            goto out_enc_err;
+        }
+    }
+
+    return true;
+
+out_enc_err:
```
Comment (skotopes) [inline firmware/targets/f7/furi_hal/furi_hal_crypto.c]: Nope

File: firmware/targets/f7/furi_hal/furi_hal_crypto.c [hunk 8]
```diff
@@ -340,3 +382,377 @@ bool furi_hal_crypto_decrypt(const uint8_t* input, uint8_t* output, size_t size)
...
+
+    /* GCM final phase */
+
+    if(!furi_hal_crypto_gcm_finish(aad_length, length, tag)) {
+        goto out_enc_err;
+    }
+
+    state = true;
+
+out_enc_err:
```
Comment (skotopes) [inline firmware/targets/f7/furi_hal/furi_hal_crypto.c]: Nope

File: firmware/targets/f7/furi_hal/furi_hal_crypto.c [hunk 9]
```diff
@@ -24,6 +24,19 @@
 #define CRYPTO_KEYSIZE_256B (AES_CR_KEYSIZE)
 #define CRYPTO_AES_CBC (AES_CR_CHMOD_0)
 
+#define CRYPTO_AES_CTR (AES_CR_CHMOD_1)
+#define CRYPTO_CTR_IV_LEN 12
```
Comment (skotopes) [inline firmware/targets/f7/furi_hal/furi_hal_crypto.c]: wrap in () and specify type

File: firmware/targets/f7/furi_hal/furi_hal_crypto.c [hunk 10]
```diff
@@ -24,6 +24,19 @@
...
+#define CRYPTO_CTR_CTR_LEN 4
+
+#define CRYPTO_AES_GCM (AES_CR_CHMOD_1 | AES_CR_CHMOD_0)
+#define CRYPTO_GCM_IV_LEN 12
+#define CRYPTO_GCM_CTR_LEN 4
+#define CRYPTO_GCM_TAG_LEN 16
+#define CRYPTO_GCM_PH_INIT 0U
```
Comment (skotopes) [inline firmware/targets/f7/furi_hal/furi_hal_crypto.c]: wrap

File: firmware/targets/f7/furi_hal/furi_hal_crypto.c (hunk 1 above)
Comment (skotopes) [reply firmware/targets/f7/furi_hal/furi_hal_crypto.c:301]: Except mode change

File: firmware/targets/f7/furi_hal/furi_hal_crypto.c [hunk 11]
```diff
@@ -276,6 +289,35 @@ bool furi_hal_crypto_store_unload_key(uint8_t slot) {
...
+    }
+
+    furi_hal_crypto_mode_init_done = false;
+    crypto_key_init((uint32_t*)key, (uint32_t*)iv);
+
+    return true;
+}
+
+bool furi_hal_crypto_unload_key(void) {
+    if(!furi_hal_bt_is_alive()) {
```
Comment (skotopes) [inline firmware/targets/f7/furi_hal/furi_hal_crypto.c]: same

File: firmware/targets/f7/furi_hal/furi_hal_crypto.c [hunk 12]
```diff
@@ -340,3 +382,377 @@ bool furi_hal_crypto_decrypt(const uint8_t* input, uint8_t* output, size_t size)
...
+
+    crypto_key_init_bswap((uint32_t*)key, (uint32_t*)iv, chaining_mode);
+
+    return true;
+}
+
+static bool wait_for_crypto(void) {
+    uint32_t countdown = CRYPTO_TIMEOUT;
+    while(!READ_BIT(AES1->SR, AES_SR_CCF)) {
+        if(LL_SYSTICK_IsActiveCounterFlag()) {
```
Comment (skotopes) [inline firmware/targets/f7/furi_hal/furi_hal_crypto.c]: use furi cortex timer

**Review by skotopes — CHANGES_REQUESTED** [review]

**Comment by skotopes** [comment]
And unit tests please ;-)

**Comment by skotopes** [comment]
Undraft when ready. 

PS: I've updated documentation and renamed couple of our methods and was ready to merge it, before I saw goto and realized that it's too complex to be merged without unit tests. Really sorry for littering your branch with it.

File: firmware/targets/f7/furi_hal/furi_hal_crypto.c (hunk 5 above)
Comment (twisted-pear) [reply firmware/targets/f7/furi_hal/furi_hal_crypto.c:552]: I do not. On some mode changes disabling CR_EN results in an invalid output according to the manual.

File: firmware/targets/f7/furi_hal/furi_hal_crypto.c (hunk 12 above)
Comment (twisted-pear) [reply firmware/targets/f7/furi_hal/furi_hal_crypto.c]: How do I do that? I copied this logic from crypto_process_block(). Can you point me to an example, where it is done correctly?

File: firmware/targets/f7/furi_hal/furi_hal_crypto.c (hunk 2 above)
Comment (twisted-pear) [reply firmware/targets/f7/furi_hal/furi_hal_crypto.c]: done in new commit

File: firmware/targets/f7/furi_hal/furi_hal_crypto.c (hunk 4 above)
Comment (twisted-pear) [reply firmware/targets/f7/furi_hal/furi_hal_crypto.c]: done in new commit

File: firmware/targets/f7/furi_hal/furi_hal_crypto.c (hunk 1 above)
Comment (twisted-pear) [reply firmware/targets/f7/furi_hal/furi_hal_crypto.c:301]: The original code also sets CR_EN once en/decryption starts. In GCM it also needs to be re-enabled when switching between certain phases.

**Comment by twisted-pear** [comment]
Added unit tests. I had some trouble getting the unit test version of the firmware running on my device, but the tests passed. I had to compile with COMPACT=1 DEBUG=0 or else fbt gave me a warning that the firmware wouldn't fit.

File: firmware/targets/f7/furi_hal/furi_hal_crypto.c (hunk 5 above)
Comment (skotopes) [reply firmware/targets/f7/furi_hal/furi_hal_crypto.c:552]: k, make sense then

File: firmware/targets/f7/furi_hal/furi_hal_crypto.c (hunk 12 above)
Comment (skotopes) [reply firmware/targets/f7/furi_hal/furi_hal_crypto.c]: Pretty much any place where you find `furi_hal_cortex_timer_get()`. It always ticking while mcu core clocked. Which is not always true for systick timer: there are some moment at early initialization stage where systick is not yet started.

File: firmware/targets/f7/furi_hal/furi_hal_crypto.c (hunk 1 above)
Comment (skotopes) [reply firmware/targets/f7/furi_hal/furi_hal_crypto.c:301]: k, make sense.

**Comment by skotopes** [comment]
@twisted-pear normally we use SD-Card for test data. But I think we can merge it as is and then move data to sd later.

**Comment by skotopes** [comment]
Also @hedger is currently working on API Version injection into defines, I think we can replace FURI_HAL_CRYPTO_ADVANCED_AVAIL with it.

Meanwhile I'll postpone this PR to next release(0.89).

**Comment by skotopes** [comment]
And that's really great work that you've done here.

**Comment by skotopes** [comment]
@twisted-pear so how about moving test data to sd?

**Comment by twisted-pear** [comment]
> @twisted-pear so how about moving test data to sd?

I'll look into it, in the coming week. I have to look at other tests that do that.

Which file format can I use to store the test vectors?

**Review by skotopes — DISMISSED** [review]

**Comment by skotopes** [comment]
Any reason why do we need `FURI_HAL_CRYPTO_ADVANCED_AVAIL`?

**Comment by twisted-pear** [comment]
> Any reason why do we need `FURI_HAL_CRYPTO_ADVANCED_AVAIL`?

I have been using it in an app I'm developing to differentiate between FW versions that support GCM (where I use furi_hal_crypto*) and those that don't (where I use a GCM implementation that I copied from somewhere) at build time. If it is possible to get and compare the API version in the C preprocessor `FURI_HAL_CRYPTO_ADVANCED_AVAIL` is not really necessary. Otherwise, it is useful to maintain compatibility across different FW versions.

**Comment by skotopes** [comment]
I'll merge this as is, but we are going to drop `FURI_HAL_CRYPTO_ADVANCED_AVAIL` closer to v1.0 replacing it with api version defines

**Review by skotopes — APPROVED** [review]

**Pull request merged**